### ISSUE TITLE
Limiting the parallelism automatically

### ISSUE THREAD

**Issue by weihanglo** [rust-lang/cargo#12912]
**tl;dr:** Introduce a simple mechanism for limiting parallelism automatically in Cargo, to avoid consuming all system resources during the compilation.

## Problem

Cargo by default uses all cores ([`std::thread::available_parallelism`](http://doc.rust-lang.org/1.73.0/std/thread/fn.available_parallelism.html))  and spawns off `rustc` or build scripts onto each core. This is not an issue when compiling on a decent machine. When working on low-end machines or large-scale codebase, developers often encounter issue like extremely high CPU loads or out-of-memory errors. To solve these problem, developers can set `--jobs` from command line, or `build.jobs` in `.cargo/config.toml` to control the maximum parallelism Cargo can use. This is not ideal because

* `build.jobs` is bound to number of core in use. It is not immediately correlated with memory usage. Executing parallel builds might lead to out of memory before any CPU throttling happens, especially when several linker invocations happens.
* `build.jobs` assigns core fairly to each unit of work (i.e. a crate build"). However, some crate builds might consume more computing resources. If those crate builds are bottlenecks of the entire compilation, we might want to throw more resources to complete them to unblock other crate builds. 
* Developers need to set `build.jobs` explicitly to control the parallelism. However, it's often a long trial-and-error to figure out a proper value. The value also varies in different environments. Not really user friendly.
* Developers might want a full control of every dependency build. `build.jobs` is too coarse-grained.

## An "ideal" approach (but not now)

There are a couple of existing proposals trying to improve the situation. Some of them want to define a weight on a certain job, or tag jobs into a group. With weights and tags, job scheduler understands whether it should allocate a job. This is pretty much the ideal solution, as it maximizes the control of parallelism for developers, and the system could be extend to the job scheduling optimization.

However, such a system requires developers to fully understand the entire compilation of their projects. For now, the data is either missing or hard to get from Cargo. To incrementally build the system, there are prerequisites:

* Cargo can monitor the resource usage of the system and each unit of work during a build.
* Cargo can persist the resource usage of each unit of work for each build.

## Start small

We should start small, and focus on monitoring the resource usage, additionally limiting the parallelism when the usage exceeds a threshold.

Some options we can do:

* Assign the maximum amount of resources that Cargo can use. This is how `build.jobs` works now. We might need an equivalent for memory usage. Something like
    ```toml
    [build.limit]
    local-memory = "3GiB" # or "95%" or "100% - 200MiB"
    ```
* Set a system threshold. Cargo won't allocate any new job and wait for the entire system usage going down, even when the usage of Cargo itself is still under the assigned maximum. 
    ```toml
    [build.limit]
    system = "3GiB" # or "95%" or "100% - 200MiB"
    cpu = "100%"
    ```

To minimize the impact of bad data points, these metrics will be sampled and averaged out within a period of time.

I entirely don't know which one we want, or both, or none.

## Library to use

* [`procfs`](https://crates.io/crates/procfs) — used by wider Rust web-dev community, via promethues and other metrics crates.
* [`sysinfo`](https://crates.io/crates/procfs) — another popular crate for inspecting system info.

Both of then introduce an excessive amount of code Cargo doesn't need at this moment.

Alternatively, we can use syscall lib directly to get these info. 

## Prior arts

* Bazel
    * `--jobs`
    * [`--local_{ram,cpu}_resources`](https://bazel.build/docs/user-manual#local-resources) to assign resources a build can use
* Buck
    * `--jobs`
    * [`link_weight`](https://buck2.build/docs/legacy/files-and-directories/dot-buckconfig/#link_weight) to config how many job a link job consumes.

* Cabel
    * `--jobs`
    * Got the same linker invocation issue <https://github.com/haskell/cabal/issues/1529>.
* CMake
    * `-j` to set max number of concurrent processes
* GitHub Actions
    * has `concurrency.group`
* Go
	* `go build -p` limits the number of programs, such as build commands or test binaries, that can be run in parallel.
    * `GOMAXPROCS` to limit the number of OS threads that can execute user-level Go code simultaneously.
* Gradle
    * `--max-workers` — like `--jobs` 
    * Has a `SharedResourceLeaseRegistry` for registering a resource with its maximum lease numbers. Like a semaphore.
    * Parallelism can be configured per-project on demand.
* make
    * `-j` to set max number of concurrent jobs
    * `--max-load` to limit the start of a new job if load average goes above the value
    * Read [Parallel](https://www.gnu.org/software/make/manual/make.html#Parallel) for more
* Ninja
    * has a [pool](https://ninja-build.org/manual.html#ref_pool) concept that user can assign some stage of build to a pool with more restricted parallelism rules.
* Nix
    * [`max-jobs`](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-max-jobs)
* sbt
    * [`tasks` are tagged](https://www.scala-sbt.org/1.x/docs/Parallel-Execution.html), and each tag get a default weight of resource restriction.

## Related issues

There are more issues regaring scheduling but I dont want to link to them here. These are issue of people trying to tell Cargo not to be that greedy.

* https://github.com/rust-lang/cargo/issues/7480
* https://github.com/rust-lang/cargo/issues/8405
* https://github.com/rust-lang/cargo/issues/8556
* https://github.com/rust-lang/cargo/issues/9157
* https://github.com/rust-lang/cargo/issues/9250
* https://github.com/rust-lang/cargo/issues/11707

And sorry I opened a new issue instead. Feel free to close and move to any existing one.

**Comment by epage** [comment]
#9250 is an interesting alternative for the CPU load aspect.  I've not done enough with nice to know how cross platform the concept is or if there are restrictions that might get in the way.

**Comment by epage** [comment]
> `local-memory = "3GiB" # or "95%" or "100% - 200MiB"`

On the surface, percentages seem nice because you don't have to worry about the exact configuration of the local system.  However, 90% of 64GB is a lot more usable of a system than 90% of 8GB.  I feel like what will be most useful is "all except".  We covered this with `--jobs` by allowing negative numbers.  We could do similar here where `-3GB` means "all but 3GB"

**Comment by epage** [comment]
With the [parallel frontend](https://github.com/rust-lang/compiler-team/issues/681) rearing its head again, we should probably consider how that affects this.

**Comment by the8472** [comment]
> We should start small, and focus on monitoring the resource usage, additionally limiting the parallelism when the usage exceeds a threshold.

On linux specifically it might be better to monitor [pressure](https://docs.kernel.org/accounting/psi.html) rather than utilization. The downside is that that's even less portable.

**Comment by clarfonthey** [comment]
FWIW, I personally was looking into a way of cross-platform measuring load average when looking for a solution to this, since it's not just Linux that would benefit from that metric. It is doable, but annoying, and I personally don't know or care enough about Windows to formulate a proper solution.

**Comment by epage** [comment]
btw doctests are another place where we can hit this (see #12916).  We should keep in mind a way to carry this information forward to those.

**Comment by clarfonthey** [comment]
Part of the reason why I mention a general solution is because, although the implementation would be complicated, the actual API probably wouldn't. Right now, things just check the number of concurrent threads the CPU can run and only let that many threads run at a time. The biggest change is that, before spawning a thread, you have to verify that both the number of threads is low enough *and* the system load is low enough. The API could be something as simple as `fn system_load() -> f64` in theory and you'd just verify that it's below the desired number.

Of course, the tricky thing is making sure that you compute that number and fetch it quickly, that it's consistent across platforms, and that you set the right load limit, which could very well depend on the system.

**Comment by weihanglo** [comment]
A zulip topic is opened for this discussion as well: <https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/parallel.20.28link.29.20jobs.20and.20OOM.20cargo.2312912>.

**Comment by luser** [comment]
> Instead of "usage", we can also leverage the concept "load average" from Unix-like, which might make more sense to manage computing resource loads.

FYI GNU make has long supported a basic version of this with its `--max-load` option: https://www.gnu.org/software/make/manual/make.html#Parallel . However, in practice it never seemed to work well, since by the time you measure that the load average is above your target, it's too late and the system is overloaded.

**Comment by the8472** [comment]
I wonder how thread priorities interact with concurrent pipe reads. I.e. do the highest-priority readers get woken up first? If so cargo could run at a higher priority than rustc or the linkers and reclaim tokens under pressure.
And rustc could release and reacquire a token every second to see if cargo wants it back.

**Comment by weihanglo** [comment]
> do the highest-priority readers get woken up first? If so cargo could run at a higher priority than rustc or the linkers and reclaim tokens under pressure.

Thanks for the suggestion. That's definitely and thing we can look into. I remember people did something similar in the past <https://github.com/rust-lang/rust/pull/67398>, but rustc turns to go the other way for parallelism story recently.

I tend to avoid a general mechanism touching every components under Rust project, which is harder to move forward. And in any case, cargo might still need a way to monitor some indicators to control the overall parallelism, from a build system perspective. They seems somehow independent and can be done separately.

**Comment by weihanglo** [comment]
> However, in practice it never seemed to work well, since by the time you measure that the load average is above your target, it's too late and the system is overloaded.

Okay, load average sounds like a lagging indicator here. @luser do you know any other indicators might help? I've done a survey for some major build tool listed in the issue descritption, but can't see any other interesting indicator they expose in the CLI interface. Or if you know any of their implementation has a fancy resource monitoring logic, please let us know. Personally I am looking for automatic way without user interference first, then we can start thinking the interface and scheduling issues.

**Comment by the8472** [comment]
Memory pressure could work to some extent because it includes page reclaims. If build processes gobble up enough ram that it forces the kernel to synchronously clean up caches or even paging that's an indication that memory reserves are running low some time before OOM conditions are reached. The question is whether it's early enough.

If a single linker job eats half the available memory but only counts as one job token then even 1 token too many can be problematic if a linker job is already running and that remaining token would be used to start another one.
Ultimately job tokens are intend to regulate core utilization, not memory utilization so there's an impedance mismatch.

Core utilization is kinda easy to regulate and predict. 1 compute-bound process/thread = 1 token.

Memory utilization is more difficult because we lack estimators for that.

Some ideas:

* tell users to add copious amounts of swap. Swap allows processes to live (even if under painful conditions), which gives us room to _react_ to swap pressure if we can't _proact_, e.g. by temporarily quiescing processes until the remaining ones can make progress without thrashing. once pressure subsides the other processes can resume work
* automatically recover from OOMs by reserving more job tokens and then restarting the child process, under the assumption that it'll succeed if it has fewer siblings
   * variation: instead of waiting for system-wide OOM conditions we can set per-process memory limits (or cgroups or whatever) and let them die earlier and then retry them later with a higher limit
* try to dial things back (by reclaiming tokens) at the earliest sign of pressure/high load/whatever. this won't help in worst-case scenarios where a single job needs all the available memory
* develop memory predictors and then have a 2nd jobserver that doles out tokens where 1 token = 256MB or something like that

**Comment by weihanglo** [comment]
Just post what I found from <https://gcc.gnu.org/wiki/DebugFission>:

> As a rule of thumb, the link job total memory requirements can be estimated at about 200% of the total size of its input files.

Might help to predict/analyze possible memory consumption for linking.

**Comment by soloturn** [comment]
here a ticket towards rust-lang, when memory consumption was unpredictable and caused OOM when compiling cosmic: https://github.com/rust-lang/rust/issues/114037 . not sure what happened. but it seems that the situation improved in august 2023.

**Comment by sunshowers** [comment]
Wanted to add that nextest also has several knobs for this:

* heavy tests, some tests take up more concurrency slots than others: https://nexte.st/book/threads-required
* test groups and mutual exclusion: https://nexte.st/book/test-groups
* more generally, profiles to control runner behavior: https://nexte.st/book/configuration.html#profiles

The context is that in nextest we also wanted to try and avoid test contention in high-core situations (e.g. https://github.com/oxidecomputer/omicron/issues/5380) -- we were looking to see if there was prior art for using an expression language to define concurrency limits, or other static/dynamic behavior. @epage kindly linked me to this thread -- thanks!

**Comment by sunshowers** [comment]
> Memory utilization is more difficult because we lack estimators for that.

A practical approach may be to record and store historical metrics.

**Comment by the8472** [comment]
https://github.com/rust-lang/rust/issues/78308 is a case of a user on a memory-constrained device running into multiple projects where the builds OOM

> Whatever the detailed reasoning is, rust seems like the only language where I'm apparently expected to learn all the compiler options first to be able to build a project at all without running out of RAM. I think that's a problem.

Part of the issue seems to be a disconnect what project maintainers configure as release profile and what a user on a low-memory device will need.

> We should start small, and focus on monitoring the resource usage, additionally limiting the parallelism when the usage exceeds a threshold.

Perhaps an even smaller start is to detect an OOM after it happened and direct users to a to dedicated documentation how to setup `config.toml` to override developer intent reduce resource consumption instead?

**Comment by ell1e** [comment]
> Perhaps an even smaller start is to detect an OOM after it happened

Do the many services like [earlyoom](https://github.com/rfjakob/earlyoom) have a standard way of reporting that? Detecting low memory might be easier.

**Comment by dpc** [comment]
Building in sandboxes (e.g. Nix), and all sorts of automation might not be compatible with recording history, detecting OOMs, etc.

From my experience, the problem is that our project became heavy, compiling it can take up to e.g. 8G of memory per job, so on machines with 8 cores, but without 64GB, it usually can't compile, and that's a rather common system configuration. And while we can explain devs how to cap it, what are we supposed to do in dockerfile's and other packaging solutions?


AFAICT, the problem is entirely with memory resource only, because it's inelastic. Sure there's a swap, etc. but in practice amount of memory puts the bound on what is possible.

It doesn't need to be perfect, simple 20/80 solution that is best effort would be way better than what is happening now.

Cases where someone is running multiple heavy things (games, other builds, etc.) are out of scope. It is expected that running out of memory will crash things. If someone is operating a build farm, runs parallel self-hosted github workers, etc. they can spend some time to tune things, and they probably already have high spec machine(s) anyway. The core thing to address is a normal unsuspecting end user that has a standard, low/mid-range machine (8 cores, 16GB ram), trying to run `docker build`, or `nix build` or `cargo build --release`, or some other packaging system and getting OOMs. Sure they could `-j2` if they are running `cargo build` directly, but if that `cargo build` is wrapped in layers of scripts and tooling, this might not be possible, and people writing such automation have no good way to help them.

The simplest solution seem to be able to tell cargo that in release build profile the expected memory usage per job is e.g. 8G of memory, so when such a configuration is detected cargo should cap the usual one per-core number of jobs to something lower that hopefully will not trigger OOMs. For this to work all is needed is:

* cargo config entries (`memory-required = "8G"`)
* cargo to be able to read total system memory on start, and use that instead of default `-j0`) - should be reasonably cheap, easy to implement, cross-platform etc.

If something more sophisticated is developed later, this simple method can just be phased out.

**Comment by emlautarom1** [comment]
I'm currently facing a similar issue with a 7950x3d and 32 GB of RAM + 16 GB of swap: building a large project forces me to use `-j4`, otherwise the system becomes extremely unstable to the point where I have to manually unplug the PC.

**Comment by the8472** [comment]
> Do the many services like [earlyoom](https://github.com/rfjakob/earlyoom) have a standard way of reporting that?

They kill the process. Few things do that, so if we see the child terminated by a SIGKILL then OOM or user action are a good guess. It can have other causes, but it might be good enough for cargo to print a hint.

**Comment by ell1e** [comment]
I think earlyoom uses SIGTERM first. However, you might not want to wait for earlyoom, it might stop something else the user cares about. Or like me, the user might stop the build earlier.

**Comment by Sieabah** [comment]
Has anyone considered the ugly solution that if you query the top processes and they're mostly pids you own (as part of the build) and when you read 100% CPU, can't you start terminating the processes? Effectively shedding the load? You don't necessarily need any of this complicated (at 100% cpu) IPC, you just need to prevent the system from killing itself. 

If not a reactive strategy then something about ramping-up the build that attempts to build a batch of targets and measure the effective "load" as it compiles. You can increase the number of batches until the desired "load" is met. My angle is that attempting to measure build complexity is a non-starter due to how many "um actually" edge cases that will be discovered. 

The goal is to create a throttled build so lower-end systems can _operate_ after running `cargo build`. I personally don't care and don't think anyone else should care if the solution kills part of the build and I wasted cycles recompiling a batch. I end wasting the entire build if my system becomes unresponsive. In fact, I lose _a lot more_ when the entire system dies, which makes any solution here better than nothing.

Probably talking into a void, but figured I'd give my thoughts since this effects my CI environment more than anything. I shouldn't have to spin up an over-provisioned dedicated metal server _exclusively for cargo_ so that I can compile something. I should be able to compile on any system and it's just a matter of time before it's done.

**Comment by dpc** [comment]
> Has anyone considered the ugly solution that if you query the top processes and they're mostly pids you own (as part of the build) and when you read 100% CPU, can't you start terminating the processes? Effectively shedding the load? You don't necessarily need any of this complicated (at 100% cpu) IPC, you just need to prevent the system from killing itself.


Many downsides comparing to [my existing proposal](https://github.com/rust-lang/cargo/issues/12912#issuecomment-3221137598).

Querying top processes and even more so cpu usage is harder, platform dependent, etc. In certain scenarios it might require calls that no permitted (e.g. some security sandbox). Then retroactively killing processes might happen too late, and also requires polling. Lastly, 100% CPU usage is normal and desired. It's only running out of memory that is the issue.

**Comment by ell1e** [comment]
So would projects need to manually set the required memory based on actual tests? For what it's worth, 8GB still seems like a fairly large amount.

**Comment by dpc** [comment]
> So would projects need to manually set the required memory based on actual tests?

I guess cargo could use a default value. Realistically we can expect a small to medium side project to require a certain amount of memory per job. I don't know ... 1 or 2 GB? So then only larger projects would need to override it.

Though I don't find it as a big of an issue because typically amount of CPUs already limits the memory usage. I don't think configurations e.g. with 8 CPU cores but only 4GB of memory are all that common.

**Comment by ell1e** [comment]
My rate of being able to build rust projects without crashing is lower than 50% so far.

**Comment by Sieabah** [comment]
@dpc You make a fair point about the complexities of process monitoring. My concern, however, isn't about finding a perfect solution, but a practical one for resource-constrained environments like CI runners or smaller cloud instances (e.g., 1 CPU, 2-4GB RAM). I shouldn't _have_ to provision a dedicated metal instance in order to compile a rust project. Coming from NodeJs/NPM land. this will only get _worse_ as more people pull in the simple dependencies from crates.io and "trivial" projects need to compile some 300-400 crates. The zed editor pulls in 1900 crates.

I disagree that targeting 100% CPU usage is always desirable. On a low-spec machine, saturating the CPU can starve the OS scheduler and I/O, leading to resource contention that actually slows down the build and can make the entire system unresponsive. I'm not sure the problem is only avoiding OOM errors; it's preventing system-wide lockups due to CPU pressure long before memory becomes an issue.

This aggressive default behavior creates a poor user experience. I've had my IDEs freeze _my entire computer_ (16 core, 4GHz & 128GB of memory) when a background build kicks off, sometimes to the point of needing a reboot. Any default build strategy that can cause scheduler thrashing or force the OS to kill processes for stability is fundamentally problematic. We're careful about always setting a threshold for channel buffer size rather than unbounded, why should the way cargo treat CPU/Memory be different in this scenario?

That's why I proposed a more adaptive, heuristic-based approach. Cargo could start with a small workload, measure the system's response, and then incrementally increase parallelism to find a sustainable throughput, always leaving a small buffer for the OS. A scrappy solution that ensures a build completes reliably is far better than a theoretically faster one that risks taking the whole system down with it. I know the platform dependent solution isn't a great start, but _anything_ is better here. Even if it only supports linux (and debian/centos/ubuntu/rhel/etc).

**Comment by ell1e** [comment]
I have never seen the same level unresponsive even on my Allwinner embedded CPU device if some processes go 100% cpu, compared to out of memory. So while neither is optimal on a slow machine, I think there's a difference (or is it just me?).

**Comment by dpc** [comment]
> I disagree that targeting 100% CPU usage is always desirable. On a low-spec machine, saturating the CPU can starve the OS scheduler and I/O, leading to resource contention that actually slows down the build and can make the entire system unresponsive. 

This is simply not true. CPUs are meant to work 100% of the time and nothing bad is happening because of it.



What you will nowadays consider "low-spec" is 10x-100x faster than what devs have been using 15 or 30 years ago, and back then 50Mhz Pentium CPU was also running compilers at full throttle and everything worked just fine. On a multi-core systems build systems other tools have been defaulting to running the amount of CPU-intense jobs equal to number of CPUs of the system forever.

The "CPU usage" is a generally useless metric. The issues you're experiencing are typically caused for one of two reasons:

1. CPU time is oversubscribed, which you will see as "load average" increasing over the number of CPUs in the system. Basically way too many jobs trying to get ahold of the CPU to run them. This also causes the CPUs to switch between jobs which thrashes caches (dcache, icache, TLB). This condition often degrades quite gracefully though, so mouse and keyboard might get some lag, but the system usually tends to stay somewhat usable.

2. If you system has swap and your build runs out of memory everything grinds to a halt because the OS has to constantly swap the memory in and out of storage, so the CPU's memory bandwidth gets capped to the IO bandwidth of the disk IO. (BTW. This will also typically cause the condition 1. to happen as well). Unlike the first case, this one is typically terrible, and the system oftentime becomes near unresponsive. I generally don't enable swap other than zswap on my machines, because I rather have things OOM and crash quickly, than look at my machines swapping while I'm unable to ssh or type in some kill command to make it fail. However enabling swap on low-spec (low-memory) machines is often necessary, just to have to get past some higher-memory usage scenarios or even browse modern websites. So low-memory systems get double-whammed.


In principle you want to avoid both conditions but  `cargo` already defaults to running number of jobs equal to the number of the CPUs. So unless something else that is heavy is running already, 1. is not an issue. The dreadful 2. is currently not being taken care of, and [my previous proposal](https://github.com/rust-lang/cargo/issues/12912#issuecomment-3221137598) is meant to precisely meant to avoid that in a relatively simple manner.

**Comment by dpc** [comment]
BTW. Not starting new jobs if loadavg is above certain (CPU core count adjusted) threshold is often used in all sorts of tooling. example [`--load` for gnu parallel](https://www.gnu.org/software/parallel/parallel_tutorial.html#limiting-the-resources). I've used it plenty of times, but it only really works well in long running scenarios with a lot of jobs with somewhat uniform distribution - for me it's typically CI jobs running tons of tests, etc.

It would not work well in Rust because the problem is typically running out of memory and that tends to happen suddenly at the end of the build when heavy top-level compilation and linking starts. As running out of memory happens suddenly, while loadavg is keeping track of historical (15s, 1m, 5m) stats of the usage. For this reason Gnu Parallel has also `--memfree` but even that would not work, because it's also looking back. Just because the machine has 8GB of free memory right now, doesn't mean that cargo will not start 2 linking jobs at the same time and each will need 5GB of memory, and the machine will grind to halt. Because of that the memory ideally needs to be accounted for in advance. cargo needs to know that "we probably need 10GB of memory to run these jobs in parallel, or we need to run them serially". And the only reliable way to get that that I can think of is to have project developers write it out somewhere in some file after they measure it in practice.

**Comment by ell1e** [comment]
I guess cargo could display a warning if a measure wasn't specified or the value was somehow determined to be possibly outdated, that might get people to do it. Unprompted I have my doubts, given job developers don't even seem to think about how the amount of parallel jobs leads to out of memory problems, which is a lower hanging fruit than measuring things.

**Comment by dpc** [comment]
It took us 3 years, 100k LoC and a handful of chunky deps to get to the point we need around 8G per job. Some default value (3G per job?) would likely eliminate the problem altogether for all projects by default.

**Comment by ell1e** [comment]
100K lines of code doesn't seem that rare for a rust project, so I feel like 3GB per job might be a little low. But I suppose it's easier to talk about defaults once the setting exists, is in practical use, and the impact can be seen hands-on.
